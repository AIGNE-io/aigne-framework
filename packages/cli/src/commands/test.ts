import { spawnSync } from "node:child_process";
import { isAbsolute, resolve } from "node:path";
import { Command } from "commander";

export function createTestCommand(): Command {
  return new Command("test")
    .description("Run tests in the specified agents directory")
    .option(
      "--url, --path <path_or_url>",
      "Path to the agents directory or URL to aigne project",
      ".",
    )
    .action(async (options: { path: string }) => {
      const { path } = options;
      const absolutePath = isAbsolute(path) ? path : resolve(process.cwd(), path);

      spawnSync("node", ["--test"], { cwd: absolutePath, stdio: "inherit" });
    })
    .showHelpAfterError(true)
    .showSuggestionAfterError(true);
}
