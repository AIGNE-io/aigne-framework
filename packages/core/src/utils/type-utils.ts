import { type ZodType, z } from "zod";

export type PromiseOrValue<T> = T | Promise<T>;

export type Nullish<T> = T | null | undefined;

export function isNil(value: unknown): value is null | undefined {
  return value === null || value === undefined;
}

export function isEmpty(obj: unknown): boolean {
  if (isNil(obj)) return true;
  if (typeof obj === "string" || Array.isArray(obj)) return obj.length === 0;
  if (typeof obj === "object") Object.keys(obj).length === 0;
  return false;
}

export function isNonNullable<T>(value: T): value is NonNullable<T> {
  return !isNil(value);
}

export function isNotEmpty<T>(arr: T[]): arr is [T, ...T[]] {
  return arr.length > 0;
}

export function orArrayToArray<T>(value?: T | T[]): T[] {
  if (isNil(value)) return [];
  return Array.isArray(value) ? value : [value];
}

export function createAccessorArray<T>(
  array: T[],
  accessor: (array: T[], name: string) => T | undefined,
): T[] & { [key: string]: T } {
  return new Proxy(array, {
    get: (t, p, r) => Reflect.get(t, p, r) ?? accessor(array, p as string),
  }) as T[] & { [key: string]: T };
}

export function checkArguments<T>(prefix: string, schema: ZodType<T>, args: T) {
  try {
    schema.parse(args, {
      errorMap: (issue, ctx) => {
        if (issue.code === "invalid_union") {
          // handle all issues that are not invalid_type
          const otherQuestions = issue.unionErrors
            .map(({ issues: [issue] }) => {
              if (issue && issue.code !== "invalid_type") {
                return issue.message || z.defaultErrorMap(issue, ctx).message;
              }
            })
            .filter(isNonNullable);
          if (otherQuestions.length) {
            return { message: otherQuestions.join(", ") };
          }

          // handle invalid_type issues
          const expected = issue.unionErrors
            .map(({ issues: [issue] }) => {
              if (issue?.code === "invalid_type") {
                return issue;
              }
            })
            .filter(isNonNullable);
          if (expected.length) {
            return {
              message: `Expected ${expected.map((i) => i.expected).join(" or ")}, received ${expected[0]?.received}`,
            };
          }
        }

        return { message: ctx.defaultError };
      },
    });
  } catch (error) {
    if (error instanceof z.ZodError) {
      const message = error.issues.map((i) => `${i.path}: ${i.message}`).join(", ");
      throw new Error(`${prefix} check arguments error: ${message}`);
    }
    throw error;
  }
}
