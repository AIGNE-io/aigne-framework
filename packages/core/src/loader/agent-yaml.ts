import { jsonSchemaToZod } from "@aigne/json-schema-to-zod";
import { nodejs } from "@aigne/platform-helpers/nodejs/index.js";
import { parse } from "yaml";
import { type ZodType, z } from "zod";
import type { AgentHooks, FunctionAgentFn, TaskRenderMode } from "../agents/agent.js";
import { AIAgentToolChoice } from "../agents/ai-agent.js";
import { type Role, roleSchema } from "../agents/chat-model.js";
import { ProcessMode, type ReflectionMode } from "../agents/team-agent.js";
import { tryOrThrow } from "../utils/type-utils.js";
import {
  camelizeSchema,
  chatModelSchema,
  defaultInputSchema,
  imageModelSchema,
  inputOutputSchema,
  optionalize,
} from "./schema.js";

export interface HooksSchema {
  priority?: AgentHooks["priority"];
  onStart?: NestAgentSchema;
  onSuccess?: NestAgentSchema;
  onError?: NestAgentSchema;
  onEnd?: NestAgentSchema;
  onSkillStart?: NestAgentSchema;
  onSkillEnd?: NestAgentSchema;
  onHandoff?: NestAgentSchema;
}

export type NestAgentSchema =
  | string
  | { url: string; defaultInput?: Record<string, any>; hooks?: HooksSchema | HooksSchema[] }
  | AgentSchema;

export interface BaseAgentSchema {
  name?: string;
  description?: string;
  model?: z.infer<typeof chatModelSchema>;
  imageModel?: z.infer<typeof imageModelSchema>;
  taskTitle?: string;
  taskRenderMode?: TaskRenderMode;
  inputSchema?: ZodType<Record<string, any>>;
  defaultInput?: Record<string, any>;
  outputSchema?: ZodType<Record<string, any>>;
  includeInputInOutput?: boolean;
  skills?: NestAgentSchema[];
  hooks?: HooksSchema | HooksSchema[];
  memory?:
    | boolean
    | {
        provider: string;
        subscribeTopic?: string[];
      };
}

export type Instructions = { role: Exclude<Role, "tool">; content: string; path: string }[];

export interface AIAgentSchema extends BaseAgentSchema {
  type: "ai";
  instructions?: Instructions;
  inputKey?: string;
  inputFileKey?: string;
  outputKey?: string;
  toolChoice?: AIAgentToolChoice;
  keepTextInToolUses?: boolean;
}

export interface ImageAgentSchema extends BaseAgentSchema {
  type: "image";
  instructions: Instructions;
  modelOptions?: Record<string, any>;
}

export interface MCPAgentSchema extends BaseAgentSchema {
  type: "mcp";
  url?: string;
  command?: string;
  args?: string[];
}

export interface TeamAgentSchema extends BaseAgentSchema {
  type: "team";
  mode?: ProcessMode;
  iterateOn?: string;
  concurrency?: number;
  iterateWithPreviousOutput?: boolean;
  includeAllStepsOutput?: boolean;
  reflection?: Omit<ReflectionMode, "reviewer"> & { reviewer: NestAgentSchema };
}

export interface TransformAgentSchema extends BaseAgentSchema {
  type: "transform";
  jsonata: string;
}

export interface FunctionAgentSchema extends BaseAgentSchema {
  type: "function";
  process: FunctionAgentFn;
}

export type AgentSchema =
  | AIAgentSchema
  | ImageAgentSchema
  | MCPAgentSchema
  | TeamAgentSchema
  | TransformAgentSchema
  | FunctionAgentSchema;

export async function parseAgentFile(path: string, data: any): Promise<AgentSchema> {
  const agentSchema: ZodType<AgentSchema> = z.lazy(() => {
    const nestAgentSchema: ZodType<NestAgentSchema> = z.lazy(() =>
      z.union([
        agentSchema,
        z.string(),
        camelizeSchema(
          z.object({
            url: z.string(),
            defaultInput: optionalize(defaultInputSchema),
            hooks: optionalize(z.union([hooksSchema, z.array(hooksSchema)])),
          }),
        ),
      ]),
    );

    const hooksSchema: ZodType<HooksSchema> = camelizeSchema(
      z.object({
        priority: optionalize(z.union([z.literal("low"), z.literal("medium"), z.literal("high")])),
        onStart: optionalize(nestAgentSchema),
        onSuccess: optionalize(nestAgentSchema),
        onError: optionalize(nestAgentSchema),
        onEnd: optionalize(nestAgentSchema),
        onSkillStart: optionalize(nestAgentSchema),
        onSkillEnd: optionalize(nestAgentSchema),
        onHandoff: optionalize(nestAgentSchema),
      }),
    );

    const baseAgentSchema = z.object({
      name: optionalize(z.string()),
      alias: optionalize(z.array(z.string())),
      description: optionalize(z.string()),
      model: optionalize(chatModelSchema),
      imageModel: optionalize(imageModelSchema),
      taskTitle: optionalize(z.string()),
      taskRenderMode: optionalize(z.union([z.literal("hide"), z.literal("collapse")])),
      inputSchema: optionalize(inputOutputSchema({ path })).transform((v) =>
        v ? jsonSchemaToZod(v) : undefined,
      ) as unknown as ZodType<BaseAgentSchema["inputSchema"]>,
      defaultInput: optionalize(defaultInputSchema),
      outputSchema: optionalize(inputOutputSchema({ path })).transform((v) =>
        v ? jsonSchemaToZod(v) : undefined,
      ) as unknown as ZodType<BaseAgentSchema["outputSchema"]>,
      includeInputInOutput: optionalize(z.boolean()),
      hooks: optionalize(z.union([hooksSchema, z.array(hooksSchema)])),
      skills: optionalize(z.array(nestAgentSchema)),
      memory: optionalize(
        z.union([
          z.boolean(),
          camelizeSchema(
            z.object({
              provider: z.string(),
              subscribeTopic: optionalize(z.array(z.string())),
            }),
          ),
        ]),
      ),
    });

    const instructionItemSchema = z.union([
      z.object({
        role: roleSchema.default("system"),
        url: z.string(),
      }),
      z.object({
        role: roleSchema.default("system"),
        content: z.string(),
      }),
    ]);

    const parseInstructionItem = async ({
      role,
      ...v
    }: z.infer<typeof instructionItemSchema>): Promise<Instructions[number]> => {
      if (role === "tool")
        throw new Error(`'tool' role is not allowed in instruction item in agent file ${path}`);

      if ("content" in v && typeof v.content === "string") {
        return { role, content: v.content, path };
      }
      if ("url" in v && typeof v.url === "string") {
        const url = nodejs.path.isAbsolute(v.url)
          ? v.url
          : nodejs.path.join(nodejs.path.dirname(path), v.url);
        return nodejs.fs.readFile(url, "utf8").then((content) => ({ role, content, path: url }));
      }
      throw new Error(
        `Invalid instruction item in agent file ${path}. Expected 'content' or 'url' property`,
      );
    };

    const instructionsSchema = z
      .union([z.string(), instructionItemSchema, z.array(instructionItemSchema)])
      .transform(async (v): Promise<Instructions> => {
        if (typeof v === "string") return [{ role: "system", content: v, path }];

        if (Array.isArray(v)) {
          return Promise.all(v.map((item) => parseInstructionItem(item)));
        }

        return [await parseInstructionItem(v)];
      }) as unknown as ZodType<Instructions>;

    return camelizeSchema(
      z.discriminatedUnion("type", [
        z
          .object({
            type: z.literal("ai"),
            instructions: optionalize(instructionsSchema),
            inputKey: optionalize(z.string()),
            outputKey: optionalize(z.string()),
            inputFileKey: optionalize(z.string()),
            outputFileKey: optionalize(z.string()),
            toolChoice: optionalize(z.nativeEnum(AIAgentToolChoice)),
            keepTextInToolUses: optionalize(z.boolean()),
            structuredStreamMode: optionalize(z.boolean()),
          })
          .extend(baseAgentSchema.shape),
        z
          .object({
            type: z.literal("image"),
            instructions: instructionsSchema,
            modelOptions: optionalize(camelizeSchema(z.record(z.any()))),
          })
          .extend(baseAgentSchema.shape),
        z
          .object({
            type: z.literal("mcp"),
            url: optionalize(z.string()),
            command: optionalize(z.string()),
            args: optionalize(z.array(z.string())),
          })
          .extend(baseAgentSchema.shape),
        z
          .object({
            type: z.literal("team"),
            mode: optionalize(z.nativeEnum(ProcessMode)),
            iterateOn: optionalize(z.string()),
            concurrency: optionalize(z.number().int().min(1)),
            iterateWithPreviousOutput: optionalize(z.boolean()),
            includeAllStepsOutput: optionalize(z.boolean()),
            reflection: camelizeSchema(
              optionalize(
                z.object({
                  reviewer: nestAgentSchema,
                  isApproved: z.string(),
                  maxIterations: optionalize(z.number().int().min(1)),
                  returnLastOnMaxIterations: optionalize(z.boolean()),
                  customErrorMessage: optionalize(z.string()),
                }),
              ),
            ),
          })
          .extend(baseAgentSchema.shape),
        z
          .object({
            type: z.literal("transform"),
            jsonata: z.string(),
          })
          .extend(baseAgentSchema.shape),
        z
          .object({
            type: z.literal("function"),
            process: z.custom<FunctionAgentFn>(),
          })
          .extend(baseAgentSchema.shape),
      ]),
    );
  });

  return agentSchema.parseAsync({
    ...data,
    model: data.model || data.chatModel || data.chat_model,
  });
}

export async function loadAgentFromYamlFile(path: string) {
  const raw = await tryOrThrow(
    () => nodejs.fs.readFile(path, "utf8"),
    (error) => new Error(`Failed to load agent definition from ${path}: ${error.message}`),
  );

  const json = tryOrThrow(
    () => parse(raw),
    (error) => new Error(`Failed to parse agent definition from ${path}: ${error.message}`),
  );

  const agent = await tryOrThrow(
    async () =>
      await parseAgentFile(path, {
        ...json,
        type: json.type ?? "ai",
        skills: json.skills ?? json.tools,
      }),

    (error) => new Error(`Failed to validate agent definition from ${path}: ${error.message}`),
  );

  return agent;
}
