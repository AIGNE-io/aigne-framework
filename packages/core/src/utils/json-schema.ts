import type { ZodType, z } from "zod";
import { zodToJsonSchema } from "zod-to-json-schema";
import type { Message } from "../agents/agent.js";
import { logger } from "./logger.js";
import { isEmpty, isRecord } from "./type-utils.js";

export function outputSchemaToResponseFormatSchema(
  agentOutput: ZodType<Message>,
): Record<string, unknown> {
  return convertNullableToOptional(
    setAdditionPropertiesDeep(zodToJsonSchema(agentOutput), false),
  )[0];
}

function setAdditionPropertiesDeep<T>(schema: T, additionalProperties: boolean): T {
  if (Array.isArray(schema)) {
    return schema.map((s) => setAdditionPropertiesDeep(s, additionalProperties)) as T;
  }
  if (schema !== null && typeof schema === "object" && !Array.isArray(schema)) {
    return Object.entries(schema).reduce(
      (acc, [key, value]) => {
        acc[key] = setAdditionPropertiesDeep(value, additionalProperties);
        if (acc.type === "object") {
          acc.additionalProperties = additionalProperties;
        }
        return acc;
      },
      {} as Record<string, unknown>,
    ) as T;
  }
  return schema;
}

export function parseJSON(json: string) {
  try {
    return JSON.parse(json);
  } catch (error) {
    logger.error("Failed to parse JSON", { json, error });
    throw error;
  }
}

/**
 * Ensure that the union array has at least 1 item.
 * NOTE: the zod union requires at least 2 items (just type definition, not runtime behavior)
 * so we need to ensure that the union has at least 1 item.
 * @param union - The union array
 * @returns The union array with at least 1 item (but the type is at least 2 items for z.union)
 */
export function ensureZodUnionArray<T extends z.ZodType>(union: T[]): [T, T, ...T[]] {
  if (!(union.length >= 1)) {
    throw new Error(`Union must have at least 1 item, but got ${union.length}`);
  }
  return union as [T, T, ...T[]];
}

export function isZodSchema(schema: ZodType): schema is ZodType {
  if (!schema || typeof schema !== "object") return false;
  return typeof schema.parse === "function" && typeof schema.safeParse === "function";
}

function convertNullableToOptional(schema: any): [typeof schema, boolean] {
  if (!isRecord(schema)) return [schema, false];

  if (schema.type === "object" && "properties" in schema && isRecord(schema.properties)) {
    const optionalProperties: string[] = [];

    const properties = Object.fromEntries(
      Object.entries(schema.properties).map(([key, value]) => {
        const [property, optional] = convertNullableToOptional(value);
        if (optional) optionalProperties.push(key);
        return [key, property];
      }),
    );
    const required =
      "required" in schema && Array.isArray(schema.required)
        ? schema.required.filter((key) => !optionalProperties.includes(key))
        : schema.required;

    return [{ ...schema, properties, required }, true];
  }

  if (schema.type === "array" && "items" in schema && isRecord(schema.items)) {
    return [{ ...schema, items: convertNullableToOptional(schema.items)[0] }, true];
  }

  if (Array.isArray(schema.type)) {
    if (schema.type.includes("null") && schema.type.length === 2) {
      return [
        {
          ...schema,
          type: schema.type.filter((t) => t !== "null")[0],
        },
        true,
      ];
    }
  }

  if (Array.isArray(schema.anyOf)) {
    const anyOf = schema.anyOf.filter(
      (i) => isRecord(i) && i.type !== "null" && !("not" in i && isEmpty(i.not)),
    );
    const optional = anyOf.length !== schema.anyOf.length;
    if (anyOf.length === 1) return [convertNullableToOptional(anyOf[0])[0], optional];
    console.log(anyOf);
  }

  return [schema, false];
}
