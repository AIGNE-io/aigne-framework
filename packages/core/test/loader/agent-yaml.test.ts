import { expect, spyOn, test } from "bun:test";
import assert from "node:assert";
import { join } from "node:path";
import {
  type Agent,
  AIAgent,
  AIGNE,
  FunctionAgent,
  ProcessMode,
  stringToAgentResponseStream,
  TeamAgent,
  TransformAgent,
} from "@aigne/core";
import { loadAgentFromYamlFile } from "@aigne/core/loader/agent-yaml.js";
import { loadAgent } from "@aigne/core/loader/index.js";
import { outputSchemaToResponseFormatSchema } from "@aigne/core/utils/json-schema.js";
import { pick } from "@aigne/core/utils/type-utils.js";
import { nodejs } from "@aigne/platform-helpers/nodejs/index.js";
import { ZodType } from "zod";
import zodToJsonSchema from "zod-to-json-schema";
import { OpenAIChatModel } from "../_mocks/mock-models.js";

test("loadAgentFromYaml should load AIAgent correctly", async () => {
  const agent = await loadAgent(join(import.meta.dirname, "../../test-agents/chat.yaml"));

  expect(agent).toBeInstanceOf(AIAgent);
  assert(agent instanceof AIAgent, "agent should be an instance of AIAgent");

  expect(agent).toEqual(
    expect.objectContaining({
      name: "chat",
      description: "Chat agent",
    }),
  );
  expect(agent.instructions.instructions).toBe(`\
You are a helpful assistant that can answer questions and provide information on a wide range of topics.
Your goal is to assist users in finding the information they need and to engage in friendly conversation.
`);

  expect(agent.skills.length).toBe(1);
  const tool = agent.skills[0];
  expect(tool).toBeInstanceOf(FunctionAgent);
  assert(tool instanceof FunctionAgent, "tool should be an instance of FunctionAgent");
  expect(tool).toEqual(
    expect.objectContaining({
      name: "evaluateJs",
      description: "This agent evaluates JavaScript code.",
    }),
  );
  expect(outputSchemaToResponseFormatSchema(tool.inputSchema)).toEqual(
    expect.objectContaining({
      type: "object",
      properties: {
        code: {
          type: "string",
          description: "JavaScript code to evaluate",
        },
      },
      required: ["code"],
    }),
  );
  expect(outputSchemaToResponseFormatSchema(tool.outputSchema)).toEqual(
    expect.objectContaining({
      type: "object",
      properties: {
        result: {
          description: "Result of the evaluated code",
        },
      },
    }),
  );
});

test("loadAgentFromYaml should error if agent.yaml file is invalid", async () => {
  spyOn(nodejs.fs, "readFile")
    .mockReturnValueOnce(Promise.reject(new Error("no such file or directory")))
    .mockReturnValueOnce(Promise.resolve("[this is not a valid yaml}"))
    .mockReturnValueOnce(Promise.resolve("name: 123"));

  expect(loadAgentFromYamlFile("./not-exist-aigne.yaml")).rejects.toThrow(
    "no such file or directory",
  );

  expect(loadAgentFromYamlFile("./invalid-aigne.yaml")).rejects.toThrow(
    "Failed to parse agent definition",
  );

  expect(loadAgentFromYamlFile("./invalid-content-aigne.yaml")).rejects.toThrow(
    "Failed to validate agent definition",
  );
});

test("loadAgentFromYaml should load mcp agent correctly", async () => {
  spyOn(nodejs.fs, "readFile")
    .mockReturnValueOnce(
      Promise.resolve(`\
type: mcp
url: http://localhost:3000/sse
`),
    )
    .mockReturnValueOnce(
      Promise.resolve(`\
type: mcp
command: npx
args: ["-y", "@modelcontextprotocol/server-filesystem", "."]
`),
    );

  expect(await loadAgentFromYamlFile("./remote-mcp.yaml")).toEqual({
    type: "mcp",
    url: "http://localhost:3000/sse",
  });

  expect(await loadAgentFromYamlFile("./local-mcp.yaml")).toEqual({
    type: "mcp",
    command: "npx",
    args: ["-y", "@modelcontextprotocol/server-filesystem", "."],
  });
});

test("loadAgentFromYaml should load TeamAgent correctly", async () => {
  const agent = await loadAgent(join(import.meta.dirname, "../../test-agents/team.yaml"));

  expect(agent).toBeInstanceOf(TeamAgent);
  assert(agent instanceof TeamAgent, "agent should be an instance of AIAgent");

  expect(agent).toEqual(
    expect.objectContaining({
      name: "test-team-agent",
      description: "Test team agent",
      mode: ProcessMode.parallel,
    }),
  );

  expect(agent.skills.length).toBe(2);
  expect(agent.iterateOn).toBe("sections");
});

test("loadAgentFromYaml should load AIAgent with prompt file correctly", async () => {
  const model = new OpenAIChatModel();

  const aigne = new AIGNE({ model });

  const agent = await loadAgent(
    join(import.meta.dirname, "../../test-agents/chat-with-prompt.yaml"),
  );

  expect(agent).toBeInstanceOf(AIAgent);
  assert(agent instanceof AIAgent, "agent should be an instance of AIAgent");

  expect(agent.instructions.instructions).toMatchSnapshot();

  const modelProcess = spyOn(model, "process").mockReturnValueOnce(
    stringToAgentResponseStream("Hello, this is a test response message"),
  );
  await aigne.invoke(agent, {
    language: "English",
  });
  expect(pick(modelProcess.mock.lastCall?.at(0) as any, "messages")).toMatchSnapshot({});
});

test("loadAgentFromYaml should load nested agent correctly", async () => {
  const agent = await loadAgent(join(import.meta.dirname, "../../test-agents/nested-agent.yaml"));

  expect(agent).toBeInstanceOf(TeamAgent);
  expect(agent.name).toMatchInlineSnapshot(`"test-nested-agent"`);
  expect(agent.description).toMatchInlineSnapshot(`"Test nested agent"`);

  const flatten = (agent: Agent): unknown => {
    return {
      name: agent.name,
      instructions: agent instanceof AIAgent ? agent.instructions?.instructions : undefined,
      inputSchema:
        agent["_inputSchema"] instanceof ZodType
          ? zodToJsonSchema(agent["_inputSchema"])
          : undefined,
      outputSchema:
        agent["_outputSchema"] instanceof ZodType
          ? zodToJsonSchema(agent["_outputSchema"])
          : undefined,
      skills: agent.skills.map((i) => flatten(i)),
    };
  };

  expect(flatten(agent)).toMatchSnapshot();
});

test("loadAgentFromYaml should load transform agent correctly", async () => {
  const agent = await loadAgent(join(import.meta.dirname, "../../test-agents/transform.yaml"));

  expect(agent).toBeInstanceOf(TransformAgent);
  assert(agent instanceof TransformAgent);
  expect(agent.name).toMatchInlineSnapshot(`"transform-agent"`);
  expect(agent.description).toMatchInlineSnapshot(`
    "A Transform Agent that processes input data using JSONata expressions.
    "
  `);
  expect(agent["jsonata"]).toMatchInlineSnapshot(`
    "{
      userId: user_id,
      userName: user_name,
      createdAt: created_at
    }
    "
  `);
});

test("loadAgentFromYaml should load external schema agent correctly", async () => {
  const agent = await loadAgent(
    join(import.meta.dirname, "../../test-agents/external-schema-agent.yaml"),
  );

  expect(zodToJsonSchema(agent.inputSchema)).toMatchInlineSnapshot(`
    {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "message": {
          "description": "Message to process",
          "type": "string",
        },
      },
      "required": [
        "message",
      ],
      "type": "object",
    }
  `);
  expect(zodToJsonSchema(agent.outputSchema)).toMatchInlineSnapshot(`
    {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "message": {
          "description": "Message to process",
          "type": "string",
        },
      },
      "required": [
        "message",
      ],
      "type": "object",
    }
  `);
});
