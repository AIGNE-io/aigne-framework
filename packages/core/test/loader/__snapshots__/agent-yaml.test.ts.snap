// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`loadAgentFromYaml should load AIAgent with prompt file correctly 1`] = `
"You are a helper agent to answer everything about chat prompt in AIGNE.

{% include "language_instruction.txt" %}
"
`;

exports[`loadAgentFromYaml should load AIAgent with prompt file correctly 2`] = `
{
  "messages": [
    {
      "content": 
"You are a helper agent to answer everything about chat prompt in AIGNE.

Output in English

"
,
      "name": undefined,
      "role": "system",
    },
  ],
}
`;

exports[`loadAgentFromYaml should load nested agent correctly 3`] = `
{
  "inputSchema": undefined,
  "instructions": undefined,
  "name": "test-nested-agent",
  "outputSchema": undefined,
  "skills": [
    {
      "inputSchema": undefined,
      "instructions": 
"You are a helpful assistant that can answer questions and provide information on a wide range of topics.
Your goal is to assist users in finding the information they need and to engage in friendly conversation.
"
,
      "name": "chat",
      "outputSchema": undefined,
      "skills": [
        {
          "inputSchema": {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "additionalProperties": false,
            "properties": {
              "code": {
                "description": "JavaScript code to evaluate",
                "type": "string",
              },
            },
            "required": [
              "code",
            ],
            "type": "object",
          },
          "instructions": undefined,
          "name": "evaluateJs",
          "outputSchema": {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "additionalProperties": false,
            "properties": {
              "result": {
                "description": "Result of the evaluated code",
              },
            },
            "type": "object",
          },
          "skills": [],
        },
      ],
    },
    {
      "inputSchema": undefined,
      "instructions": 
"You are a helper agent to answer everything about chat prompt in AIGNE.

{% include "language_instruction.txt" %}
"
,
      "name": "child",
      "outputSchema": undefined,
      "skills": [],
    },
    {
      "inputSchema": undefined,
      "instructions": undefined,
      "name": "nested-team",
      "outputSchema": undefined,
      "skills": [
        {
          "inputSchema": undefined,
          "instructions": 
"You are a helpful assistant that can answer questions and provide information on a wide range of topics.
Your goal is to assist users in finding the information they need and to engage in friendly conversation.
"
,
          "name": "chat",
          "outputSchema": undefined,
          "skills": [
            {
              "inputSchema": {
                "$schema": "http://json-schema.org/draft-07/schema#",
                "additionalProperties": false,
                "properties": {
                  "code": {
                    "description": "JavaScript code to evaluate",
                    "type": "string",
                  },
                },
                "required": [
                  "code",
                ],
                "type": "object",
              },
              "instructions": undefined,
              "name": "evaluateJs",
              "outputSchema": {
                "$schema": "http://json-schema.org/draft-07/schema#",
                "additionalProperties": false,
                "properties": {
                  "result": {
                    "description": "Result of the evaluated code",
                  },
                },
                "type": "object",
              },
              "skills": [],
            },
          ],
        },
        {
          "inputSchema": {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "additionalProperties": false,
            "properties": {
              "message": {
                "description": "Message to process",
                "type": "string",
              },
            },
            "required": [
              "message",
            ],
            "type": "object",
          },
          "instructions": 
"You are a helper agent to answer everything about chat prompt in AIGNE.

{% include "language_instruction.txt" %}
"
,
          "name": "nested-child",
          "outputSchema": {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "additionalProperties": false,
            "properties": {
              "response": {
                "description": "Response from nested child agent",
                "type": "string",
              },
            },
            "required": [
              "response",
            ],
            "type": "object",
          },
          "skills": [],
        },
      ],
    },
  ],
}
`;
